portal-kernel/test/unit/com/liferay/portal/modules/ModulesStructureTest.java

```diff
@@ -427,6 +429,63 @@ public FileVisitResult visitFile(
 		}
 	}
 
+	@Test
+	public void testScanReadmeFiles() throws IOException {
+		final PathMatcher[] excludePathMatches = {
+			_FILE_SYSTEM.getPathMatcher("glob:**/.git/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/.gradle/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/.idea/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/.m2/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/.settings/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/bin/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/build/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/classes/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/test-classes/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/test-coverage/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/test-results/**"),
+			_FILE_SYSTEM.getPathMatcher("glob:**/tmp/**")
+		};
+
+		Files.walkFileTree(
+			_modulesDirPath,
+			new SimpleFileVisitor<Path>() {
+
+				@Override
+				public FileVisitResult preVisitDirectory(
+					Path dirPath, BasicFileAttributes basicFileAttributes) {
+
+					for (PathMatcher pathMatcher : excludePathMatches) {
+						if (pathMatcher.matches(dirPath)) {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+					}
+
+					return FileVisitResult.CONTINUE;
+				}
+
+				@Override
+				public FileVisitResult visitFile(
+						Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					String fileName = String.valueOf(path.getFileName());
```

Comment:
Instead of visiting every file, I'd simply add some simple logic to the `preVisitDirectory`. If a `README.markdown` file exists in the dir, we check if `basicFileAttributes.size() == 0`... I think it's more than enough 😊

We can also exclude the path matchers, looping through directories is not too slow. Or, if we decide to add those exclusions, we should add them to all other tests as well.